### DISCUSSION TITLE
Pre-defined punishments

### DISCUSSION THREAD

**Discussion by Decollates** [gameserverapp/Platform#677] [Suggestions]
So, here me out. 

Say you're running a larger community, or a community with servers that require a more tighter, stricter ruleset for the sort of gameplay curated (such as roleplaying), or you just happen to have a large number of rulebreakers for your server. Most communities get by with a spreadsheet/google forms. Not every one of these rulebreaks deserves a ban, just a kick from the servers, or a warning, or even a temporary ban. Even for those indefinitely banned, it would be beneficial to have a logging system attached to their profile that documents the event, the staff member associated with banning, and the cause of their infraction documented as a description. 

Especially if we can have some forms of automation associated with it - for example, if a problematic player accumulates 4 active strikes, the 4th strike applied activates a task that may be set up to deliver a temporary ban. And if they still cause issues and receive a 5th strike, then their next punishment is an indefinite ban, triggered by the task that the system activates. 

I think this particular system can be adapted from the Achievements system that already exists and provides a lot of the functions I am requesting. 

![image](https://github.com/gameserverapp/Platform/assets/44717221/39081784-a7f3-4557-bd02-a0027a61ec17)

With accessing a player's profile, the individual achievements applied do have a set expiration. A punishments system can likewise be set up to issue similar tiered punishments that have set durations of "activation." This would allow communities flexibility on how long punishment strikes would last on a particular person's profile. Staff would be able to manually submit strikes to individual players, and tasks can be incorporated to automatically issue strikes on the Punishment Log via task triggers, such as forbidden words. The *reason* for such strikes could also be populated in a custom description field, with a possibility of uploading evidence (in my case, I would likely link the cloud URL of an archived discord ticket). 

![image](https://github.com/gameserverapp/Platform/assets/44717221/90e11c60-5131-43da-a937-e87f4d72e752)

Similar to the Achievements, the amount of active/inactive strikes against a player can be viewed from the `Player Information` page on their dashboard, and players can check the history of their strikes on their community website page. 

Logging for the sake of metrics can be pretty straightforward; this can be accomplished by a separate page on the dashboard, similar to the `Task Run Logs` page we have for tasks. It would also be a benefit to have discord webhooks relay activity for strikes expiring, new strikes applying, automation executing tasks, etc. 

(Sorry if this seems disjointed, I was scramming to write it within the last hour of work.)

**Comment by Decollates** [comment]
Bumping this for consideration, as I just went through a situation where the spreadsheet for player punishments was leaked to the public of our gaming communities.

**Comment by mvaessen** [comment]
I partially implemented this, because I agree that having some form of log about what happened is useful.
I haven't added a system where you can define punishments, so this is still something a human needs to decide. However, the groundwork is laid out to add automations based on historic punishments.


<img width="398" alt="Screenshot 2025-05-16 at 15 14 41" src="https://github.com/user-attachments/assets/a8ce517b-548d-402b-add2-ebc6c225f466" />
<img width="928" alt="Screenshot 2025-05-16 at 15 14 49" src="https://github.com/user-attachments/assets/2774a50a-44e4-4fa2-8b66-626e5834f8c1" />
<img width="926" alt="Screenshot 2025-05-16 at 15 14 59" src="https://github.com/user-attachments/assets/778a6047-7e76-4701-8215-a8d4a38000e0" />
